### PR TITLE
Refactor and flesh out Fn pointers

### LINKED ISSUE #507: rename the `Fn` type to `FnPointer`
The `Fn` type in chalk should be called `FnPointer` or perhaps `FunctionPointer`. Easy refactoring.

This issue has been assigned to @mental32 via [this comment](https://github.com/rust-lang/chalk/issues/507#issuecomment-641558479).

### PR DESCRIPTION
**Pull request by nathanwhit**
Closes #507.

This PR does a few things:
- Renames `Fn` -> `FnPointer`
- Adds ABI to `FnPointer`
- Adds support for unsafe functions to both `FnPointer` and `FnDef`
- Adds a variadic flag to both `FnPointer` and `FnDef` (cc [this comment](https://github.com/rust-lang/chalk/issues/507#issuecomment-658289067))

### CODE REVIEW

File: chalk-solve/src/clauses/builtin_traits/fn_family.rs [hunk 1]
```diff
@@ -138,7 +140,7 @@ pub fn add_fn_trait_program_clauses<I: Interner>(
             }
             _ => Ok(()),
         },
-        TyData::Function(fn_val) => {
+        TyData::Function(fn_val) if fn_val.safety == Safety::Safe && !fn_val.variadic => {
```
Comment (nikomatsakis) [inline chalk-solve/src/clauses/builtin_traits/fn_family.rs:143]: hmm it'd be nice if we had a more DRY way to encapsulate these two tests. We could leave for later, but basically I'm imagining if the "fn def" and "fn pointer" types embedded some common type like rustc's `FnSig` and could then invoke something like `fn_sig.implements_fn_traits()`

File: tests/test/unify.rs [hunk 2]
```diff
@@ -78,6 +78,34 @@ fn forall_equality() {
...
+        goal {
+            // Function pointers with identical ABIs should be equal.
+            for<'a, 'b> extern "Rust" fn(Ref<'a, Ref<'b, Unit>>): Eq<for<'c, 'd> extern "Rust" fn(Ref<'c, Ref<'d, Unit>>)>
+        } yields {
+            "Unique; substitution [], lifetime constraints []"
+        }
+
+        goal {
+            // Function pointers with different safety should not be equal.
+            for<'a, 'b> unsafe fn(Ref<'a, Ref<'b, Unit>>): Eq<for<'c, 'd> fn(Ref<'c, Ref<'d, Unit>>)>
```
Comment (nikomatsakis) [inline tests/test/unify.rs:98]: it feels like this test is overly complex for what it is testing. Maybe `unsafe fn(): Eq<fn()>` would be simpler?

File: tests/test/unify.rs [hunk 3]
```diff
@@ -78,6 +78,34 @@ fn forall_equality() {
...
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
+            // Function pointers with identical safety should be equal.
+            for<'a, 'b> unsafe fn(Ref<'a, Ref<'b, Unit>>): Eq<for<'c, 'd> unsafe fn(Ref<'c, Ref<'d, Unit>>)>
```
Comment (nikomatsakis) [inline tests/test/unify.rs:105]: as above

File: tests/test/unify.rs [hunk 4]
```diff
@@ -78,6 +78,34 @@ fn forall_equality() {
...
+        goal {
+            // Function pointers with different ABIs should not be equal.
+            for<'a, 'b> extern "Rust" fn(Ref<'a, Ref<'b, Unit>>): Eq<for<'c, 'd> extern "C" fn(Ref<'c, Ref<'d, Unit>>)>
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
+            // Function pointers with identical ABIs should be equal.
+            for<'a, 'b> extern "Rust" fn(Ref<'a, Ref<'b, Unit>>): Eq<for<'c, 'd> extern "Rust" fn(Ref<'c, Ref<'d, Unit>>)>
```
Comment (nikomatsakis) [inline tests/test/unify.rs:91]: can we simply, as in the previous comments?

File: tests/test/unify.rs [hunk 5]
```diff
@@ -78,6 +78,34 @@ fn forall_equality() {
             InEnvironment { environment: Env([]), goal: '!2_0: '!2_1 }, \
             InEnvironment { environment: Env([]), goal: '!2_1: '!2_0 }]"
         }
+
+        goal {
+            // Function pointers with different ABIs should not be equal.
+            for<'a, 'b> extern "Rust" fn(Ref<'a, Ref<'b, Unit>>): Eq<for<'c, 'd> extern "C" fn(Ref<'c, Ref<'d, Unit>>)>
```
Comment (nikomatsakis) [inline tests/test/unify.rs:84]: can we simply, as in the previous comments?

File: tests/test/unify.rs [hunk 6]
```diff
@@ -78,6 +78,34 @@ fn forall_equality() {
...
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
```
Comment (nikomatsakis) [inline tests/test/unify.rs:89]: shouldn't we have a test for variadics like this?

File: chalk-solve/src/clauses/builtin_traits/fn_family.rs (hunk 1 above)
Comment (nathanwhit) [reply chalk-solve/src/clauses/builtin_traits/fn_family.rs:143]: Definitely agree! I'll open an issue as a reminder.

**Comment by nathanwhit** [comment]
Addressed the comments, so this should be ready for re-review.

File: tests/test/unify.rs [hunk 7]
```diff
@@ -78,6 +78,48 @@ fn forall_equality() {
...
+            // Function pointers with identical safety should be equal.
+            unsafe fn(): Eq<unsafe fn()>
+        } yields {
+            "Unique; substitution [], lifetime constraints []"
+        }
+
+        goal {
+            // Variadic function pointers should not be equal to non-variadic fn pointers.
+            variadic fn(): Eq<fn()>
```
Comment (nikomatsakis) [inline tests/test/unify.rs:112]: Side note: the actual Rust syntax is `fn(...)` or something like that, maybe we should use that instead?

**Comment by jackh726** [comment]
Okay this is as good as any of a PR to test out bors.

@bors r+

**Comment by jackh726** [comment]
Welp that didn't work

**Pull request merged**